packages/yew/src/dom_bundle/btag/mod.rs

```diff
@@ -252,9 +253,35 @@ impl VTag {
                 .create_element_ns(namespace, tag)
                 .expect("can't create namespaced element for vtag")
         } else {
-            document()
-                .create_element(tag)
-                .expect("can't create element for vtag")
+            thread_local! {
+                static CACHED_ELEMENTS: RefCell<Vec<(String, Element)>> = Default::default();
```

Comment:
I'm not sure I can follow your argument for using a linear search here. Since it's caching by tag, we might even be able to fine-tune the hashing used to avoid collisions, but even without this, a `HashMap` would be less surprising. The "usual" website uses between 40-100 different tagNames.

I ran `new Set(Array.from(document.querySelectorAll("*")).map(e => e.tagName)) ` on [the top 50 list](https://en.wikipedia.org/wiki/List_of_most_visited_websites). A few outliers are explained by the liberal use of custom elements on the google pages especially. The above also counts `html`, `body` and a few other elements that probably do not appear in the app itself, but I think expecting the average yew app to use 30 different elements at least is reasonable. I don't see the linear search being faster than a lookup in the map, and the memory overhead it most likely negligible.

Would you try this with a `HashMap` with a default capacity of, say 32?

Reply:
Sure, let me give a bit more of my reasoning, since we should start from the assumption that a HashMap is the right call here.
1. For any given comparison between `O(n)` and `O(1)` it's good to keep in mind that if `n` is relatively small and `1` is relatively large, `n` may actually be more efficient. For example, if we only have 2 items in the Vec, it will obviously be cheaper to do a linear search than to look it up in a HashMap. If we have 10,000 items, it will obviously be cheaper to look it up in the HashMap. 
2. I made the guess that n = 30 is somewhere in the "not a significant difference" range. I agree this is important to actually test rather than making an assumption.
3. Relative to the cost of DOM rendering itself these differences are likely minimal.
4. Binary size: Yew already has `Vec<Element>` in it so this doesn't add meaningful binary size. It doesn't (afaict) have a `HashMap<String, Element>` anywhere, so this is a new data structure to be monomorphized and included in the binary. 

I did just do a HashMap version with capacity 32 as you suggested. Here are the benchmark results
<img width="265" alt="Screenshot 2023-03-18 at 5 37 05 PM" src="https://user-images.githubusercontent.com/286622/226141649-8d005b78-ef95-4b56-8caf-b50fb9410caa.png">
<img width="264" alt="Screenshot 2023-03-18 at 5 37 16 PM" src="https://user-images.githubusercontent.com/286622/226141701-a05448d2-acf9-47a2-893d-99dab3b01515.png">

I'm pleased to say you're right and I'm wrong here, in that even at this small n the HashMap is winning. You can see it when creating 1000 elements if you average "create 1000" and "append 1000", and a much bigger difference at 10,000. Of course on this particular run this wasn't enough to swamp the general statistical noise so the Vec approach was "faster" overall but this is not significant.

Note however that the HashMap version adds another 4kb to the WASM binary size.

So that's the tradeoff to consider in using a HashMap instead: even faster element creation that may or may not be measurable vs. 4kb in the binary.

Reply:
Is 4KB (I assume you meant _bytes with a B_ and not _bits_) really that much difference when WASM binary is highly compressible and is streamed to the client? I'm not sure. In any sizeable application, the binary size can be in MBs. 4KB is merely a drop in the bucket for that